ffi/webp.lua

```diff
@@ -0,0 +1,139 @@
+--[[--
+Module for WebP decoding.
+
+This is a LuaJIT FFI wrapper for libwebp.
+
+@module ffi.webp
+]]
+
+local ffi = require("ffi")
+local BB = require("ffi/blitbuffer")
+
+require("ffi/libwebp_h")
+
+-- local libwebp = ffi.load("libs/libwebp." .. (ffi.os == "OSX" and "7.dylib" or "so.7"))
+-- We only need the stuff provided by libwebpdemux.so (which itself uses libwebp.so)
+local libwebpdemux = ffi.load("libs/libwebpdemux." .. (ffi.os == "OSX" and "2.dylib" or "so.2"))
+
+local Webp = {
+    webp_decoder = nil,
+    webp_image = nil,
+    width = nil,
+    height = nil,
+    components = 4, -- always RGB32
+    nb_frames = nil,
+    cur_frame = nil,
+}
+
+function Webp:new(o)
+    o = o or {}
+    setmetatable(o, self)
+    self.__index = self
+    return o
+end
+
+function Webp.fromFile(filename)
+    local fh = io.open(filename, "rb")
+    assert(fh, "couldn't open WebP file")
+    local data = fh:read("*a")
+    fh:close()
+    return Webp.fromData(data)
+end
+
+function Webp.fromData(data, size)
+    local webp_data = ffi.new("WebPData[1]")
+    webp_data[0].bytes = ffi.cast("const unsigned char*", data)
+    webp_data[0].size = size or #data
+
+    -- Note: WebPAnimDecoderOptionsInit() and WebPAnimDecoderNew() are inline functions in demux.h.
+    -- We need to use their ...Internal() versions.
+    local dec_options = ffi.new("WebPAnimDecoderOptions[1]")
+    local ret = libwebpdemux.WebPAnimDecoderOptionsInitInternal(dec_options, libwebpdemux.WEBP_DEMUX_ABI_VERSION);
+    assert(ret ~= 0, "libwebp WebPAnimDecoderOptionsInit() failed.")
+    dec_options[0].color_mode = libwebpdemux.MODE_RGBA
+
+    local decoder = libwebpdemux.WebPAnimDecoderNewInternal(webp_data, dec_options, libwebpdemux.WEBP_DEMUX_ABI_VERSION)
+    -- check for nil for "NULL in case of parsing error, invalid option or memory error"
+    assert(decoder ~= nil, "libwebp WebPAnimDecoderNew() failed (parsing or memory error).")
+
+    local anim_info = ffi.new("WebPAnimInfo[1]")
+    ret = libwebpdemux.WebPAnimDecoderGetInfo(decoder, anim_info)
+    assert(ret ~= 0, "libwebp WebPAnimDecoderGetInfo() failed.")
+
+    if anim_info[0].frame_count > 1 and type(data) ~= "string" then
+        -- When called from CreDocument:getImageFromPosition(), data is a "userdata" wrapping
+        -- a buffer malloc()'ed by cre.cpp, which is freed as soon as getImageFromPosition()
+        -- returns, as we usually don't need it after having rendered the image data to
+        -- a blitbuffer.
+        -- With animated multiframes images, we keep the decoder object: GifLib is fine with
+        -- data being gone, but libwebp is not.
+        -- We could have renderImageData() return a flag to state that the data should not
+        -- be free()d, but it would complexify the return values and their handling.
+        -- So, do simple: make a copy of the external buffer into an interned Lua string,
```

Comment:
nit: `keep it simple` ;).

Reply:
Heh, at first I thought you were talking about the code.